File: src/algorithms/rappor.js
const crypto = require('crypto')
const struct = require('python-struct')
const _ = require('lodash')
const zipf = require('./zipf')
const lasso = require('./lasso')
const random = require('random')
const matrix = require('dstructs-matrix')

const NUM_BLOOM_BITS = 8
const NUM_HASH = 2
const NUM_COHORT = 1
const NUM_REPORTS = 500000
const NUM_ITEM = 10
const P = 0.5
const Q = 0.75
const F = 0.5
const ZIPF_DISTRIBUTION = 1.01

function makingDictionary(numItem, numCohort, numHash, numBloomBits) {

  const ret = []
  _.range(numItem).forEach((item) => {
    _.range(1, numCohort+1).forEach((cohort) => {
      const userItem = `v${item}`
      const value = struct.pack('>L', cohort) + userItem
      const digest = crypto.createHash('md5').update(value).digest()

      let line = `${userItem} ${cohort} `
      line += _.range(numHash).map((i) => {
        return digest[i] % numBloomBits
      }).join(' ')
      ret.push(line)
    })
  })
  return ret
}

function sumBits(bit1, bit2, numBloomBits) {
  const ret = _.range(numBloomBits).map((i) => {
    return parseInt(bit1[i], 10)+parseInt(bit2[i], 10)
  })
  return ret
}

function pmf(x, distribution) {
  return zipf.pmf(x, distribution)
}

function ComputeWeight(numItem, distribution) {
  return _.range(1, numItem+1).map((item) => {
    return pmf(parseFloat(item), distribution)
  })
}

function SampleZipf(numReports, numItem, distribution) {
  const weights = ComputeWeight(numItem, distribution)
  const ret = []
  const itemFreq = []
  let sumWeights = weights.reduce((a, b) => a + b)
  _.range(numItem).forEach((item) => {
    const itemP = Math.min(Math.max(parseFloat(weights[item]/sumWeights), 0.0), 1.0)
    const itemList = numReports
    const rndSampleArray = random.binomial(itemList, itemP)
    const rndSample = rndSampleArray()
    numReports -= rndSample
    sumWeights -= weights[item]
    itemFreq.push(rndSample)
    ret.push(item + ' ' + rndSample + ' ' + itemP)
  })
  return itemFreq
}

function shuffle(a) {
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
}

function makingInput(numReports, numItem, numCohort, distribution) {
  const itemFreq = SampleZipf(numReports, numItem, distribution)
  console.log('itemFreq', itemFreq)
  const userData = {}
  let tmpList = []
  _.range(itemFreq.length).forEach((item) => {
    _.range(itemFreq[item]).forEach((i) => {
      tmpList.push(item)
    })
  })
  console.log('tmpList size', tmpList.length)
  tmpList = shuffle(tmpList)
  console.log('tmpList size', tmpList.length)

  const ret = []
  _.range(tmpList.length).forEach((user) => {
    userData[user] = {}
    userData[user].item = `v${tmpList[user]}`
    userData[user].cohort = random.int(1, numCohort)
    ret.push(`${userData[user].item},${userData[user].cohort}`)
  })
  return userData
}

function bloomFilter(userData, numBloomBits, numHash) {
  const value = struct.pack('>L', userData.cohort) + userData.item
  const digest = crypto.createHash('md5').update(value).digest()
  while (1) {
    if (digest.length >= numHash) { break }
  }
  return _.range(numHash).map(i => (digest[i] % numBloomBits))
}

function Rappor(userData, numBloomBits, numHash, p, q, f) {
  const blmFilter = bloomFilter(userData, numBloomBits, numHash)
  let str1 = ''
  _.range(numBloomBits).forEach((i) => {
    if (blmFilter.indexOf(i) === -1) {
      str1 += '0'
    } else {
      str1 += '1'
    }
  })
  const fakeBlmFilter = PermanentRandomized(blmFilter, f)
  const report = InstaneousRandomized(fakeBlmFilter, p, q)
  return report
}

function PermanentRandomized(userBlmFilter, f) {
  let r = 0
  let str1 = ''
  _.range(NUM_BLOOM_BITS).forEach((i) => {
    const ran = random.float()
    if (ran < parseFloat(0.5 * f)) {
      str1 += '1'
    } else if (ran < 2 * parseFloat(0.5 * f)) {
      str1 += '0'
    } else {
      if (userBlmFilter.indexOf(i) !== -1) {
        str1 += '1'
      } else {
        str1 += '0'
      }
    }
  })
  r = parseInt(str1, 2)
  return r
}

function InstaneousRandomized(userFakeBlmFilter, p, q) {
  const report = []
  _.range(NUM_BLOOM_BITS).forEach((bitNum) => {
    const ran = random.float()
    if (userFakeBlmFilter & (1 << bitNum)) {
      if (ran < q) {
        report.push('1')
      } else {
        report.push('0')
      }
    } else {
      if (ran < p) {
        report.push('1')
      } else {
        report.push('0')
      }
    }
  })
  report.reverse()
  return report.join('')
}

function single_perturb(value) {
  const userData = {};
  userData.cohort = random.int(1, NUM_COHORT);
  userData.item = `v${value}`
  const rappor = Rappor(userData, NUM_BLOOM_BITS, NUM_HASH, P, Q, F);

  //return { cohort: userData.cohort, perturbed: rappor };
  return { perturbed: rappor };
}
if (typeof window !== 'undefined') {
  window.rappor_perturb = single_perturb;
}

if (typeof module !== 'undefined' && module.exports) {
  module.exports = {
    makingDictionary,
    sumBits,
  }
}

/*
const vectorC = {}
const vectorT = {}

const yMat = _.fill(_.range(NUM_COHORT * NUM_BLOOM_BITS), 0)
const xMat = _.range(NUM_COHORT * NUM_BLOOM_BITS).map(() => _.fill(_.range(NUM_ITEM), 0))


const cohortReport = {}
const result = {}
const resultMean = {}

_.range(NUM_ITEM).forEach((item) => {
  const value = `v${item}`
  result[value] = {}
  resultMean[value] = {}
  _.range(1, NUM_COHORT+1).forEach((cohort) => {
    result[value][cohort] = []
    resultMean[value][cohort] = 0.0
  })
})

_.range(1, NUM_COHORT+1).forEach((cohort) => {
  cohortReport[cohort] = 0
  vectorC[cohort] = _.fill(_.range(NUM_BLOOM_BITS), 0)
  vectorT[cohort] = _.fill(_.range(NUM_BLOOM_BITS), 0)
})

const dict = makingDictionary(NUM_ITEM, NUM_COHORT, NUM_HASH, NUM_BLOOM_BITS)
const userData = makingInput(NUM_REPORTS, NUM_ITEM, NUM_COHORT, ZIPF_DISTRIBUTION)

_.range(Object.keys(userData).length).forEach((i) => {
  cohortReport[userData[i].cohort] += 1
  vectorC[userData[i].cohort] = sumBits(vectorC[userData[i].cohort], Rappor(userData[i], NUM_BLOOM_BITS, NUM_HASH, P, Q, F), NUM_BLOOM_BITS)
})

_.range(1, NUM_COHORT+1).forEach((i) => {
  _.range(NUM_BLOOM_BITS).forEach((j) => {
    vectorT[i][j] = (parseFloat(vectorC[i][j]) - (parseFloat(P) + (0.5 * F * Q) - (0.5 * F * P)) * parseFloat(cohortReport[i])) / ((1-F) * (Q-P))
    yMat[(i-1) * NUM_BLOOM_BITS + j] = vectorT[i][j]
  })
})

let cnt = 0
let lineCnt = 0

dict.forEach((line) => {
  lineCnt += 1
  line = line.trim()
  const words = line.split(' ')
  const item = words[0].slice(1)
  const cohort = parseInt(words[1])
  let meanHash = 0.0
  _.range(2, NUM_HASH+2).forEach((index) => {
    const bloomBitsIndex = parseInt(words[index])
    i = parseInt((cohort-1) * NUM_BLOOM_BITS + bloomBitsIndex)
    j = parseInt(item)
    xMat[i][j] = 1
  })
})

console.log(xMat)
console.log(yMat)
const xMatObj = matrix(_.flatten(xMat), [NUM_COHORT * NUM_BLOOM_BITS, NUM_ITEM])
const yMatObj = new Float64Array(yMat)
const out = lasso(xMatObj, yMatObj, 1.0)
console.log(out)
*/

File: src/algorithms/rappor.py
#-*- coding:utf-8 -*-
import hashlib
import sys
import struct
import numpy as np
from random import *
from scipy.stats import zipf
from sklearn import linear_model
from sklearn.feature_selection import SelectFromModel
import sys
import json

num_bloom_bits = 8 
num_hash = 2
num_cohort = 1 
num_reports = 500000
num_item = 10
p = 0.5
q = 0.75
f = 0.5
zipf_distribution = 1.01



# Making hash dictionary
# 각 cohort 마다
# 각 item의  bloom filter에서의 hash 위치를 기록해 놓는다.
# (같은 아이템도 cohort 별로 다른 hash 값을 갖는다)
def making_dictionary(num_item,num_cohort,num_hash,num_bloom_bits):

    fw = open('dictionary.txt','w')
    for item in range(num_item):
        for cohort in range(1,num_cohort+1):
            user_item = 'v'+ str(item)                  
            value = struct.pack('>L',cohort) + user_item            # cohort 값과 item에 따라서 다른 value가 나온다. 
            md5 = hashlib.md5(value)                                 
            digest = md5.digest()                                   # value값으로 hash 위치 지정
            fw.write(user_item + ' ' + str(cohort))
            for i in xrange(num_hash):
                fw.write(' ' + str(ord(digest[i])%num_bloom_bits))
            fw.write('\n')


# Sum of bits
# 두 벡터 값을 더하기 위한 함수
def sum_bits(bit1, bit2,num_bloom_bits):
    ret = []
    for i in range(0,num_bloom_bits):
        ret.append(int(bit1[i])+int(bit2[i]))

    return ret


# zipf distribution's probability mass function
# 지프 분포에 따라서 값을 도출한다
def pmf(x,distribution):

    ret = zipf.pmf(x,distribution)
    return ret


# Compute item's Weights
# 아이템의 weight를 계산한다
def ComputeWeight(num_item,distribution):

    weights = []
    for item in range(1,num_item+1):
        item = float(item)
        weights.append(pmf(item,distribution))
    
    return weights


# Sampling frequency
# global 변수인 zipf 분포 값을 이용하여 각 item의 frequency를 계산한다.
def Sample_zipf(num_reports,num_item,distribution):
    
    weights = ComputeWeight(num_item,distribution)
    item_freq = []
    sum_weights = sum(weights)
    fw = open('true_value.txt','w')
    for item in range(num_item):
        item_p = min(max(weights[item]/sum_weights,0),1)
        item_list = num_reports
        rnd_sample_array = np.random.binomial(item_list,item_p,1)
        rnd_sample = rnd_sample_array[0]
        num_reports -= rnd_sample
        sum_weights -= weights[item]
        item_freq.append(rnd_sample)
        fw.write(str(item) + ' ' + str(rnd_sample)+ ' ' + str(item_p) +'\n')
    
    return item_freq

# Making Input
# global 변수인 zipf 분포 값을 이용하여 input을 생성한다.
def making_input(num_reports, num_item, num_cohort, distribution):

    fw = open('input.csv','w')
    item_freq = Sample_zipf(num_reports, num_item, distribution)        # line 74: 각 item들의 frequency를 계산한다.
    user_data = {}
    tmp_list = []
    for item in range(0,len(item_freq)):
        for i in range(0,item_freq[item]):
            tmp_list.append(item)                                       
    """
    cnt = 0;
    cnt2 = 0;
    cnt3 = 0;
    cnt4 = 0;
    for i in range(0,len(tmp_list)):
        if tmp_list[i] == 0:
            cnt += 1;
        if tmp_list[i] == 1:
            cnt2 += 1;
        if tmp_list[i] == 2:
            cnt3 += 1;
        if tmp_list[i] == 3:
            cnt4 += 1;
    print cnt
    print cnt2
    print cnt3
    print cnt4
    """
    shuffle(tmp_list)

    for user in range(0,len(tmp_list)):
        user_data[user]={}
        user_data[user]['item'] = 'v'+str(tmp_list[user])
        user_data[user]['cohort'] = randint(1,num_cohort)
        fw.write(str(user_data[user]['item'])+','+str(user_data[user]['cohort'])+'\n')

    fw.close()
    return user_data
    


# Bloom Filter
# user의 item으로 bloom_filter에서의 hash위치를 찾는다.
def bloom_filter(user_data,num_bloom_bits,num_hash):

    value = struct.pack('>L',user_data['cohort']) + user_data['item']   # cohort 값과 item에 따라서 다른 값(value)이 나온다.
    md5 = hashlib.md5(value)                                            # value값으로 hash의 위치 지정
    while(1):
        digest = md5.digest()                                           
        if len(digest) >=  num_hash:
            break
    
    return [ord(digest[i]) % num_bloom_bits for i in xrange(num_hash)]


# Whole RAPPOR process
def Rappor(user_data,num_bloom_bits,num_hash,p,q,f):

    blm_filter = bloom_filter(user_data,num_bloom_bits,num_hash)        # user의 item을 hash에 통과시켜 bloom_filter 의위치 지정
    str1 = ""                                                            
    for i in range(0,num_bloom_bits):
        if i not  in blm_filter:
            str1+='0'
        else:
            str1+='1'
    fake_blm_filter = Permanent_Randomized(blm_filter,f)                # Permanent randomize를 실행한다.
    
    report = Instaneous_Randomized(fake_blm_filter,p,q)                 # Instaneous randomize를 실행한다.
    return report



# Phase1: Bloom Filter -> perturbed data
# item을 hash에 통과 시켜 만들어진 각 bloom filter를randomize한다.
# 각bit의 값 Bi를
#{ 1 , with probability 1/2f }
#{ 0 , with probability 1/2f }
#{ Bi, with probability 1-f  }
# 로 coin tossing 해준다.
def Permanent_Randomized(user_blm_filter,f):

    r = 0
    str1 = ""
    for i in range(0,num_bloom_bits):
        ran = random()
        if ran < float(0.5*f):              # Bi = 1 , with probability 1/2f 
            str1 += "1"
        elif ran < 2*float(0.5*f):          # Bi = 0 , with probability 1/2f 
            str1 += "0"
        else:                               # Bi = Bi , with probability 1-f 
            if i in user_blm_filter:
                str1 += "1"
            else:
                str1 += "0"
    r = int(str1,2)
    return r


# Phase2: perturbed data -> perturbed data
# Permanent randomize 된 값을 instanos randomize한다.
# 이는 중복된 query에 대해서 보호하기 위함
# 각 bit의 값 Bi를
#{ Bi , with probability q }
#{ NOT Bi , with probability p }
def Instaneous_Randomized(user_fake_blm_filter,p,q):
    report = []
    for bit_num in range(0,num_bloom_bits):
        ran = random()
        if user_fake_blm_filter & ( 1 << bit_num):
            if ran < q:                                 # Bi = Bi , with probability q 
                report.append('1')
            else:
                report.append('0')
        else:                                           # Bi = Bi의 반전 , with probability p 
            if ran < p:
                report.append('1')
            else:
                report.append('0')
    report.reverse() 
    return ''.join(report)
    

if __name__=="__main__":

    vector_c = {}
    vector_t = {}
    y_mat = np.zeros((num_cohort*num_bloom_bits))
    x_mat = np.zeros((num_cohort*num_bloom_bits,num_item))

    cohort_report = {}
    result = {}
    result_mean = {}
    for item in range(0,num_item):
        value = 'v'+str(item)
        result[value]= {}
        result_mean[value] = {}
        for cohort in range(1,num_cohort+1):
            result[value][cohort] = []
            result_mean[value][cohort] = 0.0

    for cohort in range(1,num_cohort+1):
        cohort_report[cohort] = 0
        vector_c[cohort] = [0] * num_bloom_bits
        vector_t[cohort] = [0] * num_bloom_bits
    
                                                                            # 여기까지는 변수initialize


    making_dictionary(num_item,num_cohort,num_hash,num_bloom_bits)                      # 각 cohort별 item의 hash 위치를 저장해 놓는다.
    user_data = making_input(num_reports, num_item, num_cohort, zipf_distribution)      # 각 user가 가진 item으로bloom filter의 hash 위치를 찾는다.

    with open('user_data.json', 'w') as fpp:
        json.dump(user_data, fpp)

    for i in range(0,len(user_data)):
        cohort_report[user_data[i]['cohort']] += 1                                      # cohort별 report수를 카운트
        #print user_data[i]['cohort'],cohort_report[user_data[i]['cohort']]
        vector_c[user_data[i]['cohort']] = sum_bits(vector_c[user_data[i]['cohort']],Rappor(user_data[i],num_bloom_bits,num_hash,p,q,f),num_bloom_bits)
                                                                                        # 각 유저별로 Randomize 함수인 Rappor를 수행시켜 해당 벡터값들을 모두 합산한다=> vector_c
    for i in range(1,num_cohort+1):
        for j in range(0,num_bloom_bits):
            vector_t[i][j] = (float(vector_c[i][j]) - (float(p) + (0.5*f*q) - (0.5*f*p))*float(cohort_report[i])) / ((1-f)*(q-p))
                                                                                        # 합산된 vector 값은biased하기 때문에 unbiased하게 조정해준다=> vector_t 
            y_mat[(i-1)*num_bloom_bits+j] = vector_t[i][j]                              # vector_t를  { 가로 : cohort 수 * bloom bit수, 세로 : 1 }
                                                                                        # 사이즈를   가지는 matrix로 변환한다     
                                                                                        # => y_mat
    
    print('vector_c', vector_c)
    print('vector_t', vector_t)
    
                                                                                        # line 242 에서 만들어진 dictionary로 { 가로 : item 수 , 세로 : cohort 수 *  bloom_bit 수}
    fr = open('dictionary.txt','r')                                                     # 사이즈를 가지는(각 cohort별 각 item의 hash 위치) matrix를 생성한다       
    cnt = 0                                                                             # => x_mat
    line_cnt = 0                                                                        
    lines = fr.readlines()
    for line in lines:
        line_cnt += 1
      
        line = line.strip()
        words = line.split(' ')
        
        item = words[0][1:]
        cohort = int(words[1])
        mean_hash = 0.0
        for index in range(2,num_hash+2):
            bloom_bits_index = int(words[index])
            i = int((cohort-1)*num_bloom_bits+bloom_bits_index)
            j = int(item)
            x_mat[i][j] = 1                                                              

                                                                                        # x_mat과 y_mat을 LASSO regression을 통해 각 item별 coefficient를 구한다 => frequency 
                                                                                         
    print(x_mat)
    print(y_mat)
    model = linear_model.Lasso(alpha=0.5,normalize=True,fit_intercept = False)          # (Rasso)linear regression 
    x_mat = x_mat.tolist()
    y_mat = y_mat.tolist()
    a_model = model.fit(x_mat,y_mat)    
    print model.coef_                                                                   # 각 item별 frequency를 출력
    print('len', len(model.coef_))


